### PR TITLE
fix(terminal): handler cursor shape correctly

### LINKED ISSUE #3681: :terminal cursor doesn't match $TERM cursor
It looks like NeoVim is manually drawing the cursor in `:terminal` windows, rather than letting the external terminal (in my case, urxvt) draw the cursor. This results in some irksome inconsistency in how my cursor looks in `:terminal` windows vs. how it looks everywhere else, including other buffers in NeoVim. Specifically, I have my terminal cursor set to blink. Everywhere else in NeoVim, my cursor blinks, but in `:terminal` windows, it doesn't.

NeoVim really ought to be letting the external terminal draw the cursor.

### PR DESCRIPTION
**Pull request by glepnir**
use $TERM cursor in terminal buffer and restore our cursor in normal buffer

Fix #3681

### CODE REVIEW

**Comment by glepnir** [comment]
just an early draft. There is still some work to be done. currently working on other PRs. wait for a while.

File: src/nvim/api/ui_events.in.h [hunk 1]
```diff
@@ -35,6 +35,10 @@ void set_title(String title)
   FUNC_API_SINCE(3);
 void set_icon(String icon)
   FUNC_API_SINCE(3);
+void set_cursor(void)
+  FUNC_API_SINCE(3);
+void reset_cursor(void)
+  FUNC_API_SINCE(3);
```
Comment (justinmk) [inline src/nvim/api/ui_events.in.h:41]: `mode_info_set` already specifies the cursor styles, and presumably the UI can react to `mode_change`. Why do we need new UI events?

**Comment by WieeRd** [comment]
@glepnir Looks like the progress of this PR is stale.
If you don't have time to work on this right now, do you mind giving us some vague explanation on the idea you had in mind?
It'll help people - including me - experiment around the cursor fix and perhaps make some progress.

**Comment by gpanders** [comment]
See https://github.com/neovim/neovim/pull/31562

**Pull request closed without merging**